parser_library/src/workspaces/file_manager_impl.cpp

```diff
@@ -14,56 +14,329 @@
+        , m_lines(that.m_lines)
+        , m_fm(that.m_fm)
+        , m_lsp_version(that.m_lsp_version)
+    {}
+
+    ~mapped_file()
+    {
+        std::lock_guard guard(m_fm.files_mutex);
+        if (m_it != m_fm.m_files.end())
+            m_fm.m_files.erase(m_it);
+    }
+
+    // Inherited via file
+    const file_location& get_location() const override { return m_location; }
+    const std::string& get_text() const override { return m_text; }
+    bool get_lsp_editing() const override { return m_editing_self_reference != nullptr; }
+    version_t get_version() const override { return m_version; }
+    bool error() const override { return m_error.has_value(); }
+
+    bool up_to_date() const override
+    {
+        std::lock_guard guard(m_fm.files_mutex);
+        return m_it != m_fm.m_files.end();
+    }
+
+    std::optional<std::string_view> get_text_or_error() const
+    {
+        if (m_error.has_value())
+            return std::nullopt;
+        else
+            return m_text;
+    }
+};
+
+class : public external_file_reader
+{
+    std::optional<std::string> load_text(const utils::resource::resource_location& document_loc) final
+    {
+        return utils::resource::load_text(document_loc);
+    }
+} default_reader;
+
+file_manager_impl::file_manager_impl()
+    : file_manager_impl(default_reader)
+{}
+
+file_manager_impl::file_manager_impl(external_file_reader& file_reader)
+    : m_file_reader(&file_reader)
+{}
+
+file_manager_impl::~file_manager_impl()
+{
+    std::shared_ptr<mapped_file> to_release;
+    std::shared_ptr<mapped_file>* release = &to_release;
+
+    std::unique_lock lock(files_mutex);
+
+    for (auto& [_, mf] : m_files)
+    {
+        if (mf.file->m_editing_self_reference)
+        {
+            *release = std::move(mf.file->m_editing_self_reference);
+            release = &mf.file->m_editing_self_reference;
+        }
+    }
+
+    lock.unlock();
+
+    while (to_release)
+    {
+        auto next = std::move(to_release->m_editing_self_reference);
+        std::swap(next, to_release);
+    }
+}
+
+template<typename... Args>
+std::shared_ptr<file_manager_impl::mapped_file> file_manager_impl::make_mapped_file(Args&&... args)
+{
+    auto result = std::make_shared<mapped_file>(std::forward<Args>(args)...);
+    result->m_self = result;
+    return result;
+}
+
+std::shared_ptr<file> file_manager_impl::add_file(const file_location& file_name)
+{
+    std::unique_lock lock(files_mutex);
+
+    if (auto result = try_obtaining_file_unsafe(file_name, nullptr))
+        return result;
+
+    lock.unlock();
+
+    auto loaded_text = m_file_reader->load_text(file_name);
+
+    lock.lock();
+
+    if (auto result = try_obtaining_file_unsafe(file_name, &loaded_text))
+        return result;
+
+    auto result = loaded_text.has_value() ? make_mapped_file(file_name, *this, loaded_text.value())
+                                          : make_mapped_file(file_name, *this, mapped_file::file_error());
+
```

Comment:
```suggestion

    lock.unlock();
    
    auto result = loaded_text.has_value() ? make_mapped_file(file_name, *this, loaded_text.value())
                                          : make_mapped_file(file_name, *this, mapped_file::file_error());
                                          
    lock.lock();

```

?

Reply:
I would leave it like this then.

However, I'm still not sure about the double checking via `try_obtaining_file_unsafe()` before and after `m_file_reader->load_text(file_name)`. What's the scenario when the first check would fail, how often could that happen and how realistic it could be?

Reply:
That's a classic race where 2 threads try to open the same file, the solution I decided to go for is to use the content from the thread that finishes the reading first.

Reply:
Seems legit! 👍